### PR TITLE
Add CustomEnvironmentFacetBuilder class

### PR DESCRIPTION
**Pull request by Anirudh181001**
### Problem
The main idea behind this pull request is to enable a way to capture custom environment variables from a spark integration. The flow defined here, as applicable to our use case is to have a CUSTOM_VARS="VAR1,VAR2,VAR3..." environment variable that indicates which environment variables should be captured. Then, we would want to capture those variables, ie VAR1, VAR2, VAR3, ... from the spark clusters.

Closes: #1419

### Solution

The solution proposed extends the EnvironmentFacets class, to build a CustomEnvironmentFacetBuilder class. This class then is used to get the values of the environment variables and pass on to the startEvent json.

- [ ] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [x] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [ ] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [ ] Your changes are accompanied by tests (_if relevant_)
- [ ] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)
- [ ] You've added a [header](https://github.com/OpenLineage/OpenLineage/tree/main/.github/header_templates.md) to source files (_if relevant_)

----
SPDX-License-Identifier: Apache-2.0\
Copyright 2018-2023 contributors to the OpenLineage project

### CODE REVIEW

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+ * {@link CustomFacetBuilder} that generates a {@link EnvironmentFacet} when using OpenLineage on
+ * Databricks.
+ */
+@Slf4j
+public class CustomEnvironmentFacetBuilder
+    extends CustomFacetBuilder<SparkListenerJobStart, EnvironmentFacet> {
+  private Map<String, Object> envProperties;
+
+  public static boolean isCustomEnvironmentVarCollectionEnabled() {
+    return System.getenv().containsKey("CUSTOM_VARS");
```
Comment (mobuchowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java:32]: I think longer form `CUSTOM_ENV_VARIABLES` would be preferred, even if more annoying to type. Or even `CUSTOM_OPENLINEAGE_ENV_VARIABLES`. Not strongly attached to this though.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  }
+
+  @Override
+  protected void build(
+      SparkListenerJobStart event, BiConsumer<String, ? super EnvironmentFacet> consumer) {
+    consumer.accept(
+        "environment-properties", new EnvironmentFacet(getCustomEnvironmentalAttributes(event)));
```
Comment (mobuchowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java:39]: I wonder if this won't override another `DatabricksEnvironmentFacetBuilder` that emits the same facet. Maybe there should be some merging there, or those facet builders be merged.

@pawel-big-lebowski what do you think?

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java (hunk 1 above)
Comment (mobuchowski) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java:32]: Also, I think this is something that should also be specified as Spark config, as there is rest of our configuration settings.

For example, as `spark.openlineage.facets.custom_env_vars`

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  private Map<String, Object> getCustomEnvironmentalAttributes(SparkListenerJobStart jobStart) {
+    envProperties = new HashMap<>();
+    // These are useful properties to extract if they are available
+    String customVars = System.getenv().get("CUSTOM_VARS");
+    List<String> customPropertiesToExtract = Arrays.asList(customVars.split(","));
+
+    customPropertiesToExtract.stream()
+        .forEach(
+            (p) -> {
+              envProperties.put(toCamelCase(p, false, new char[] {'_'}), System.getenv().get(p));
```
Comment (mobuchowski) [inline integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java:102]: Why do we need to convert those to camelCase?

**Review by mobuchowski — CHANGES_REQUESTED** [review]
Hey, most of this looks good. I've commented at few things in the code; also to get rid of test values you need to 

- rebase your PR on main, which will fix the Python test failures
- run `./gradlew spotlessApply` - make sure you're using Java 8 here, it works differently on different Java versions
- sign your commits https://github.com/OpenLineage/OpenLineage/pull/1545/checks?check_run_id=11056802342

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java (hunk 3 above)
Comment (Anirudh181001) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java:102]: Hmm yeah this is optional...more so for our specific use case, but yeah I will remove this part from the class.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java (hunk 2 above)
Comment (Anirudh181001) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java:39]: There will be an override I think...I will edit this class to merge with the custom vars instead.

File: integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java (hunk 1 above)
Comment (Anirudh181001) [reply integration/spark/shared/src/main/java/io/openlineage/spark/agent/facets/builder/CustomEnvironmentFacetBuilder.java]: Yeap that makes sense. I am now passing it in as a spark config.

**Comment by pawel-big-lebowski** [comment]
The code looks good to me. 
Would it be possible to create a test (similar to this one: https://github.com/OpenLineage/OpenLineage/blob/main/integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/DeltaDataSourceTest.java) which verifies presence of env variable in RunEvent? 

That's the best guarantee the feature will not be broken in future.

**Comment by Anirudh181001** [comment]
> The code looks good to me. Would it be possible to create a test (similar to this one: https://github.com/OpenLineage/OpenLineage/blob/main/integration/spark/app/src/test/java/io/openlineage/spark/agent/lifecycle/DeltaDataSourceTest.java) which verifies presence of env variable in RunEvent?
> 
> That's the best guarantee the feature will not be broken in future.

Hi @pawel-big-lebowski Thanks for the suggestion! The DeltaDataSourceTest is failing the assertions on local, and seems like it is not able to read any lineageEvent in the test. Is this expected behaviour?

**Review by mobuchowski — APPROVED** [review]

**Comment by mobuchowski** [comment]
Thanks for the contribution @Anirudh181001

File: client/java/src/main/java/io/openlineage/client/transports/FacetsConfig.java [hunk 4]
```diff
@@ -21,4 +21,8 @@ public class FacetsConfig {
   @Getter
   @JsonProperty("disabled")
   private String[] disabledFacets;
+
+  @Getter
+  @JsonProperty("custom_environment_variables")
```
Comment (pawel-big-lebowski) [inline client/java/src/main/java/io/openlineage/client/transports/FacetsConfig.java:26]: @Anirudh181001 I know it's a kind of a late remark, but would you be able to include description of this parameter in Spark integration docs: https://github.com/OpenLineage/OpenLineage/tree/main/integration/spark#general  ?

**Pull request merged**